### ISSUE TITLE
Inline CSV parsing fails with CRLF (Windows) in map.xml

### ISSUE THREAD

**Issue by BergWerkGIS** [mapnik/mapnik#3320]
Running `node-mapnik@latest` tests against `mapnik@latest` these failed:

```
  1) mapnik.VectorTile.composite should support compositing tiles that were just rendered to sync:
     Uncaught Error: first argument must be either a layer name (string) or layer index (integer)
      at Error (native)
      at c:\mb\windows-builds-64\packages\node-mapnik\test\vector-tile.composite.test.js:207:32
      at c:\mb\windows-builds-64\packages\node-mapnik\test\vector-tile.composite.test.js:67:16

  2) mapnik.VectorTile.composite should support compositing tiles that were just rendered to sync (reencode):
      Uncaught AssertionError: 0 == 49
      + expected - actual
      -0
      +49
      at c:\mb\windows-builds-64\packages\node-mapnik\test\vector-tile.composite.test.js:237:20
      at c:\mb\windows-builds-64\packages\node-mapnik\test\vector-tile.composite.test.js:67:16

  3) mapnik.VectorTile.composite should support compositing tiles that were just rendered to async:
      Uncaught AssertionError: 0 == 49
      + expected - actual
      -0
      +49
      at c:\mb\windows-builds-64\packages\node-mapnik\test\vector-tile.composite.test.js:341:20
      at c:\mb\windows-builds-64\packages\node-mapnik\test\vector-tile.composite.test.js:67:16

  4) mapnik.VectorTile.composite should support compositing tiles that were just rendered to async (reencode):
      Uncaught AssertionError: 0 == 49
      + expected - actual
      -0
      +49
      at c:\mb\windows-builds-64\packages\node-mapnik\test\vector-tile.composite.test.js:371:20
      at c:\mb\windows-builds-64\packages\node-mapnik\test\vector-tile.composite.test.js:67:16
```

during execution of the tests some `Mapnik LOG`s are shown:

<pre>
  mapnik.VectorTile.composite
    √ should fail to composite due to bad parameters
<b>Mapnik LOG> 2016-02-26 11:00:29: CSV Plugin: # of columns(2) > # of headers(1) parsed at line: 2</b>
<b>Mapnik LOG> 2016-02-26 11:00:29: CSV Plugin: # of columns(2) > # of headers(1) parsed at line: 3</b>
    1) should support compositing tiles that were just rendered to sync
<b>Mapnik LOG> 2016-02-26 11:00:29: CSV Plugin: # of columns(2) > # of headers(1) parsed at line: 2</b>
<b>Mapnik LOG> 2016-02-26 11:00:29: CSV Plugin: # of columns(2) > # of headers(1) parsed at line: 3</b>
    2) should support compositing tiles that were just rendered to sync (reencode)
    √ should support compositing tiles and clipping to max_extent (reencode) (118ms)
<b>Mapnik LOG> 2016-02-26 11:00:29: CSV Plugin: # of columns(2) > # of headers(1) parsed at line: 2</b>
<b>Mapnik LOG> 2016-02-26 11:00:29: CSV Plugin: # of columns(2) > # of headers(1) parsed at line: 3</b>
    3) should support compositing tiles that were just rendered to async
<b>Mapnik LOG> 2016-02-26 11:00:29: CSV Plugin: # of columns(2) > # of headers(1) parsed at line: 2</b>
<b>Mapnik LOG> 2016-02-26 11:00:29: CSV Plugin: # of columns(2) > # of headers(1) parsed at line: 3</b>
</pre>

The failing tests reference https://github.com/mapnik/node-mapnik/blob/master/test/data/vector_tile/compositing/layers/lines.xml

If `git` is installed with default settings on Windows it changes the line feed characters of text files to the system default `CRLF`.

_Thus `lines.xml` has `CRLF` line endings which seems to screw the inline CSV logic.
Changing the line ending to `LF` made the errors disappear._

Would be great, if the inline CSV logic would also take care of Windows line endings.
Thinking about supporting `map.xml`s and CSVs that have been created natively on Windows.

----

BTW: gettings _hundreds_ of other CSV errors during `node-mapnik` tests, but they don't trigger any test failures.

e.g.:
```
Mapnik LOG> 2016-02-26 11:16:19: CSV Plugin: # of columns(10) > # of headers(9) parsed at line: 736
```

**Comment by artemp** [comment]
@BergWerkGIS - both inline and file based CSV use the same logic so this is unexpected.

**Comment by artemp** [comment]
> gettings hundreds of other CSV errors during node-mapnik tests, but they don't trigger any test failures.

Those are not errors just warnings ? 

@BergWerkGIS ^

**Comment by lightmare** [comment]
The problem might be that `autodetect_csv_flavour` and `getline_csv` try to figure out *one* newline character, while they should consume `\r?\n` (not a lone `\r`).

**Comment by artemp** [comment]
Ok this sneaked in during `csv_grammar` re-factor . The fix applied in 3ee109f46b56415f122ef8614cd8e5d3be879da8 but I'm confused why our test didn't catch this. Perhaps git settings .. 

/cc @springmeyer @lightmare

**Comment by artemp** [comment]
@BergWerkGIS - could you test 3ee109f is working on windows, cheers.

**Comment by BergWerkGIS** [comment]
> Those are not errors just warnings ?

Sorry, of course, `Mapnik LOG` messages like `CSV Plugin: # of columns(10) > # of headers(9)` and not errors or failed tests.

> but I'm confused why our test didn't catch this. Perhaps git settings ..

Yeah, they do catch it, but we set `IGNORE_FAILEDTEST=1` last year, when there was a known (harmless) failing test to get the build server run thru and somehow it was not reset after the release was out.

Working on getting tests running on AppVeyor.

**Comment by artemp** [comment]
@BergWerkGIS - does master after https://github.com/mapnik/mapnik/commit/a4d7f03bb659e885dd74d5e78347c00c4a34872c works?

**Comment by BergWerkGIS** [comment]
@artemp no CSV related `Mapnik LOG`s for me locally (AppVeyor is setup to clone to `LF`).

First iteration of getting Python tests working on AppVeyor:
https://ci.appveyor.com/project/Mapbox/mapnik/build/1.0.706#L1901

**Comment by springmeyer** [comment]
@BergWerkGIS I'm having trouble parsing your comment. Are you saying `no` to @artemp asking if latest code works? Or are you saying `there are no warnings` anymore so things are working correctly now? Can you summarize overall where we are at here and next steps?

**Comment by BergWerkGIS** [comment]
* no `Mapnik LOG` warnings anymore locally -> CSV parsing works again (with `CRLF` line endings. A git default install converts text file `LF` line endings to `CRLF` during cloning).
* AppVeyor always worked (clones as `LF`)

Next steps:
* enable Python tests on AppVeyor
* enable other tests on AppVeyor
* maybe, if we don't run into the time limit, enable benchmarks on AppVeyor

Friction:
* Python builds were always picking up the wrong Python install (32bit) on AppVeyor (worked locally) -> solved
* PostgreSQL service doesn't seem to get started on AppVeyor, even when specified in the `appveyor.yml` -> solved (our scripts start it now)
* Creating `postgis_template` -> still wip

**Comment by springmeyer** [comment]
Thanks for the extra details @BergWerkGIS. One thing I'm concerned about is the changing of files that git does by default. We want to make sure that the line endings in the CSV testcases are unmodified. This is important so that we are testing that we support all line ending types across all operating systems. So, on OS X and Linux we should be able to support CSV with windows line endings. And on Windows we should also be able to support windows line endings and unix. Can you advise on what git settings we should be consistently using such that:

  - CSV testcases are unmodified by git
  - We ensure that the tests on appveyor start failing if this support regresses (again).

In short: the key problem I see is that appveyor did not previously catch this regression and I want to make sure we do what we need to make sure it does in the future.

**Comment by lightmare** [comment]
Preferably with [gitattributes](http://git-scm.com/docs/gitattributes) `*.csv -text` rather than git settings.

**Comment by BergWerkGIS** [comment]
@springmeyer 

> So, on OS X and Linux we should be able to support CSV with windows line endings. And on Windows we should also be able to support windows line endings and unix.

Since we are covering all platforms by CI my initial thought would be to let git convert everything to the native settings of the corresponding OS.
We would then get failures on the specific platform if mapnik is not able to handle the native format.

Second thought:
Have three files with OSX, Linux and Windows line endings and tell git to not mangle just those.
So we get coverage on all platform for all types of line endings.

---

@lightmare 

> Preferably with gitattributes *.csv -text rather than git settings.

yes, that's how we solved this problem before, eg. with libosmium.

> edit: thinking about it, no file in test-data should ever be tinkered with by git on checkin/checkout, so: `* -text`

Don't agree :100: %.
I think text files should be in the native format of the OS.
Software that is cross platform should be able to handle that, otherwise it isn't really cross platform.

Pinning line endings to `LF` and tests to always expect `LF` might hide some serious flaws.

**Comment by lightmare** [comment]
@BergWerkGIS 

| Pinning line endings to LF and tests to always expect LF might hide some serious flaws.

What I meant was that test-data should contain files of all flavours, so that all are tested on each platform.

I wouldn't exclude individual files because when some files undergo automagic conversion and others not, it's rather surprising. Three is not enough. The tests need to do more than line-counting, they need to ensure different line endings work in grammar constructs -- escaped newlines, quoted strings and such. And then there are files containing weird mixtures:
```
find test-data -name '*newline*'
test-data/csv/fails/needs_headers_one_line_no_newline.csv
test-data/csv/mac_newlines_with_unix_inline.csv
test-data/csv/windows_newlines_with_unix_inline_escaped.csv
test-data/csv/windows_newlines.csv
test-data/csv/trailing_newline_mac.csv
test-data/csv/mac_newlines_with_unix_inline_escaped.csv
test-data/csv/mac_newlines.csv
test-data/csv/unix_newlines_with_windows_inline_escaped.csv
test-data/csv/trailing_newline_win.csv
test-data/csv/windows_newlines_with_unix_inline.csv
test-data/csv/unix_newlines_with_windows_inline.csv
```

Plus, having varying line endings in the entire dataset (csv, xml, json, ...) would test their parsing more thoroughly than a few hand-crafted dummy files with a few lines.

**Comment by BergWerkGIS** [comment]
> I wouldn't exclude individual files because when some files undergo automagic conversion and others not, it's rather surprising. Three is not enough

Exactly, that's why I opt for not changing git defaults to always get OS specific line endings.

> And then there are files containing weird mixtures

Yes, and then those with special formatting that should not be changed by git. 

> Have three files for ...

Alright, `Have three batches of files for ...`.

Nevertheless, while tests are important and a _lot_ of time should be spent on them, I think we should also keep effort and outcome in balance, because covering every single possible configuration/file type/use case/... that will possibly ever come our way upfront is just not feasible.

**Comment by artemp** [comment]
`<BDFL>` :smiling_imp: :smiling_imp: :smiling_imp: 
##### Thank you everyone for contributing to this thread. Here is what I'm going to enforce from now on! 
* I added `.gitattributes` in `test/data` repository with `* -text` 
* I added some extra test files with `DOS` (EOL => CRLF) and `MAC` (EOL => CR) encodings  

I'm OK with having three batches for all tests, perhaps creating `dos`, `mac` directories for the relevant data files.

*The above change is very important because previously some bugs could go unnoticed for weeks!!! 
For example - https://github.com/mapnik/mapnik/issues/3359*

> Nevertheless, while tests are important and a lot of time should be spent on them, I think we should also keep effort and outcome in balance, because covering every single possible configuration/file type/use case/... that will possibly ever come our way upfront is just not feasible.

@BergWerkGIS ^  I hear you but we're talking about only three commonly used EOL (right, I'm not sure who is using `CR` but we're supporting it, so let's have it too! 


`</BDFL>`

**Comment by lightmare** [comment]
> @artemp I'm OK with having three batches for all tests...

In case this refers to:

> @lightmare test-data should contain files of all flavours, so that all are tested on each platform.

I didn't mean tripling all tests. I imagined that `* -text` alone would naturally create a mix of tests with various line endings over time -- so in addition to explicit line-endings-tests, other tests would also test line endings as a side effect, as each new test will have line endings given by its creator.

**Comment by artemp** [comment]
Concluded in f28e96de1e2eae0d88393ff4bae6e4885d9b608f and 8bf0d18da9c716fe33905893d81818d5632070a9
